File: boa/interpreter.py
import io
import sys
from stdlib.boa_math import Math
from stdlib.boa_time import Time
from stdlib.boa_arr  import Arr
from stdlib.boa_map  import Map
from environment import Environment
from expr import ExprVisitor, Literal, Var
from statement import StmtVisitor, ArrayStmt
from token_types import TokenType

def assert_eq(a, b, message):
    assert a == b, message

def assert_boa(a, message):
    assert a, message


class Callable:
    def __init__(self, func, arity):
        self.func  = func
        self.arity = arity

    def call(self, args):
        return self.func(*args)

class CallableFunc:
    def __init__(self, decl, arity):
        self.decl = decl
        self.arity = arity

    def call(self, interpreter, args):
        env = Environment(interpreter.globals)
        for param, arg in zip(self.decl.params, args):
            env.define(param.lexeme, arg)
        try:
            interpreter.execute_block(self.decl.body, env)
        except ReturnException as r:
            return r.return_statement
        return None

class Interpreter(StmtVisitor, ExprVisitor):
    def __init__(self, debug_mode=False):
        self.statements = []
        self.globals = Environment()
        self.env = self.globals
        self.globals.define("assert_eq", Callable(assert_eq, 3))
        self.globals.define("assert", Callable(assert_boa, 2))
        self.output = io.StringIO()
        self.saved_stdout = sys.stdout
        self.debug_mode = debug_mode

    def __enter__(self):
        if self.debug_mode:
            sys.stdout = self.output
        return self

    def __exit__(self):
        if self.debug_mode:
            sys.stdout = self.saved_stdout

    def get_output(self):
        return self.output.getvalue()

    def interpret(self, statements):
        result = []
        for statement in statements:
            r = self.execute_statement(statement)
            result.append(r)
        return result

    def execute_statement(self, stmt):
        return stmt.accept(self)

    def parse_to_float(self, value):
        try:
            if isinstance(value, Var):
                value = self.env.get(value.ident.lexeme)
            elif isinstance(value, Literal):
                value = value.value
            return round(float(value), 3), True
        except Exception:
            return value, False

    def visit_expression_statement(self, stmt):
        # return None In reality expression statements produce no values 
        # (We return for testing)
        return self.evaluate(stmt.expression)

    def visit_print_statement(self, stmt):
        val = self.evaluate(stmt.expression)
        if (isinstance(val, bool) or val) and self.parse_to_float(val)[1]:
            try:
                print(round(val, 3))
            except Exception:
                print(val)
        elif val == None:
            print ("nil")
        else:
            print(val)
        return None

    def visit_if_statement(self, ifstmt):
        pred = self.evaluate(ifstmt.predicate)
        else_block = ifstmt.else_block
        if pred:
            self.visit_block_statement(ifstmt.block)
        if else_block and not pred:
            self.visit_block_statement(ifstmt.else_block)

    def visit_while_statement(self, whilestmt):
        while self.evaluate(whilestmt.predicate) == True:
            self.visit_block_statement(whilestmt.block)

    def visit_loop_statement(self, forstmt):
        self.visit_var_statement(forstmt.start)
        while self.evaluate(forstmt.predicate) == True:
            self.visit_block_statement(forstmt.block)
            self.evaluate(forstmt.incrementer)

    def evaluate_array(self, array):
        evaluated = []
        for element in array:
            if isinstance(element, list):
                evaluated.append(self.evaluate_array(element))
            elif hasattr(element, "lexeme"):
                if element.type == TokenType.IDENTIFIER:
                    evaluated.append(self.env.get(element.lexeme))
                else:
                    evaluated.append(element.lexeme)
            else:
                evaluated.append(self.evaluate(element))
        return evaluated

    def visit_var_statement(self, stmt):
        from functools import reduce
        identifier = stmt.ident
        if stmt.expression != None:
            if isinstance(stmt.expression, ArrayStmt):
                val = stmt.expression
                if val.elements != None:
                    val = self.evaluate_array(val.elements)
                else:
                    index = [int(self.evaluate(x)) for x in val.index]
                    val = reduce(lambda acc, idx: acc[idx], index, 
                        self.env.get(stmt.expression.ident.lexeme))
                self.env.define(identifier.lexeme, val)
            else:
                val = self.evaluate(stmt.expression)
                self.env.define(identifier.lexeme, val)
        else:
            self.env.define(identifier.lexeme, None)
        return self.env.get(identifier.lexeme)

    def visit_block_statement(self, block):
        return self.execute_block(block, Environment(self.env))

    def visit_logical_expression(self, logicalexpr):
        left = self.evaluate(logicalexpr.left)

        if logicalexpr.op.type == TokenType.OR:
            if self.is_truthy(left):
                return left
        else:
            if not self.is_truthy(left):
                return left

        return self.evaluate(logicalexpr.right)

    def execute_block(self, block, env):
        prev = self.env
        self.env = env
        res = []
        for statement in block.statements:
            res.append(self.evaluate(statement))
        self.env = prev
        return res

    def visit_assign_expression(self, expr):
        self.env.assign(expr.ident, self.evaluate(expr.value))
        return self.env.get(expr.ident.lexeme)

    def visit_binary_expression(self, expr):
        left  = self.evaluate(expr.left)
        right = self.evaluate(expr.right)
        left,  l_parsed = self.parse_to_float(left)
        right, r_parsed = self.parse_to_float(right)

        if l_parsed and r_parsed:
            if expr.op.type == TokenType.PLUS:
                return left + right
            if expr.op.type == TokenType.MINUS:
                return left - right
            if expr.op.type == TokenType.SLASH:
                return left / right
            if expr.op.type == TokenType.STAR:
                return left * right
            if expr.op.type == TokenType.EQUAL_EQUAL:
                return left == right
            if expr.op.type == TokenType.BANG_EQUAL:
                return left != right
            if expr.op.type == TokenType.GREATER:
                return left > right
            if expr.op.type == TokenType.LESS:
                return left < right
            if expr.op.type == TokenType.LESS_EQUAL:
                return left <= right
            if expr.op.type == TokenType.GREATER_EQUAL:
                return left >= right
        else:
            if isinstance(left, list) and isinstance(right, list) and expr.op.type == TokenType.PLUS:
                return left + right
            if expr.op.type == TokenType.PLUS:
                l = str(left).strip('"')
                r = str(right).strip('"')

                l = l.replace("None", "nil")
                r = r.replace("None", "nil")

                f = l + r
                return f'"{f}"'
            if expr.op.type == TokenType.BANG_EQUAL:
                return left != right
            if expr.op.type == TokenType.EQUAL_EQUAL:
                return left == right
        return None

    def visit_literal_expression(self, expr):
        return expr.value

    def visit_grouping_expression(self, expr):
        return self.evaluate(expr.expression)

    def visit_unary_expression(self, expr):
        right = self.evaluate(expr.right)
        if expr.op.type == TokenType.MINUS:
            return float(right) * -1
        if expr.op.type == TokenType.BANG:
            return not self.is_truthy(right)
        return None

    def visit_var_expression(self, expr):
        return self.env.get(expr.ident.lexeme)

    def visit_call_expression(self, expr):
        if len(expr.args) >= 128:
            raise Exception("To many arguments have been provided")
        func = self.evaluate(expr.callee)
        if func is None or not (isinstance(func, Callable) or isinstance(func,
                                                                         CallableFunc)):
            raise Exception(f"Error with defined func")
        if func.arity != len(expr.args):
            raise Exception(f"Error with the defined number of args")
        args = []
        for arg in expr.args:
            args.append(self.evaluate(arg))
        if isinstance(func, CallableFunc):
            import copy
            return func.call(copy.copy(self), args)
        return func.call(args)
    
    def visit_func_statement(self, visitor):
        func = CallableFunc(visitor, len(visitor.params))
        self.env.define(visitor.token_name.lexeme, func)
        return None

    def visit_return_statement(self, visitor):
        val = None
        if visitor.value != None:
            val = self.evaluate(visitor.value)
        raise ReturnException(val)

    def visit_array_statement(self, visitor):
        from functools import reduce
        try:
            getter = self.env.get(visitor.ident.lexeme)
            if isinstance(getter, list):
                if isinstance(visitor.index, Var):
                    index = self.env.get(visitor.index.ident.lexeme)
                else:
                    index = [int(self.evaluate(x)) for x in visitor.index]
                return reduce(lambda acc, idx: acc[idx], index, getter)
            elif isinstance(getter, dict):
                if isinstance(visitor.index, Var):
                    index = self.env.get(visitor.index.ident.lexeme)
                else:
                    index = []
                    for x in visitor.index:
                        x = self.evaluate(x)
                        if type(x) == float:
                            index.append(int(x))
                        else:
                            index.append(x)
                return reduce(lambda acc, idx: acc[idx], index, getter)
        except Exception:
            raise IndexError("Array does not contain index ", visitor.index)

    def visit_array_assign_statement(self, visitor):
        value = self.evaluate(visitor.value)
        array = self.env.get(visitor.ident.lexeme)
        indices = [int(self.evaluate(x)) for x in visitor.index]
        for idx in indices[:-1]:
            array = array[idx]
        # Assign the value at the final index
        array[indices[-1]] = value
        return None

    def visit_hash_map_statement(self, visitor):
        keys   = [self.evaluate(k) for k in visitor.keys]
        values = [self.evaluate(v) for v in visitor.values]
        if None in keys:
            raise Exception("nil values are not allowed as keys in hashmap")
        return dict(zip(keys, values))

    def visit_array_expression(self, visitor):
        return [self.evaluate(v) for v in visitor.elements]
    
    def visit_import_statement(self, visitor):
        allowed_libs = ["math", "arr", "time", "map"]
        if visitor.lib_name.lexeme not in allowed_libs:
            raise Exception(f"import {visitor.lib_name.lexeme} not found")
        if visitor.lib_name.lexeme == "map":
            map = Map()
            self.globals.define("map.insert", Callable(map.insert, 3))
            self.globals.define("map.keys", Callable(map.keys, 1))
            self.globals.define("map.values", Callable(map.values, 1))
            self.globals.define("map.get", Callable(map.get, 2))
        if visitor.lib_name.lexeme == "math":
            math = Math()
            self.globals.define("math.pi", Callable(math.pi, 0))
            self.globals.define("math.pow", Callable(math.pow, 2))
            self.globals.define("math.factorial", Callable(math.factorial, 1))
            self.globals.define("math.ceil", Callable(math.ceil, 1))
            self.globals.define("math.floor", Callable(math.floor, 1))
            self.globals.define("math.is_prime", Callable(math.is_prime, 1))
            self.globals.define("math.min", Callable(math.min, 2))
            self.globals.define("math.max", Callable(math.max, 2))
            self.globals.define("math.sqrt", Callable(math.sqrt, 1))
            self.globals.define("math.abs", Callable(math.abs, 1))
            self.globals.define("math.random_f", Callable(math.random_f, 0))
        if visitor.lib_name.lexeme == "time":
            time = Time()
            self.globals.define("time.sleep", Callable(time.sleep, 1))
            self.globals.define("time.clock", Callable(time.clock, 0))
        if visitor.lib_name.lexeme == "arr":
            arr = Arr()
            self.globals.define("arr.length", Callable(arr.length, 1))
            self.globals.define("arr.append", Callable(arr.append, 2))
            self.globals.define("arr.pop", Callable(arr.pop, 1))
            self.globals.define("arr.modify", Callable(arr.modify, 3))
        return

    def is_truthy(self, expr):
        if expr == None:
            return False
        if isinstance(expr, bool):
            return bool(expr)
        return True
    
    def evaluate(self, expr):
        return expr.accept(self)

class ReturnException(Exception):
    def __init__(self, return_statement):
        self.return_statement = return_statement

File: boa_vm/interpreter.go
package main

import (
  "fmt"
  "reflect"
  "time"
)
  

type Interpreter struct {
  env        *Env
  statements []Statement
}

type ReturnValue struct {
  value interface{}
}

type Clock struct {
  arity int32
}

func (clock *Clock) call(interpreter *Interpreter, args []interface{}){
  t := time.Now().Unix()
  fmt.Println(t)
}

func NewInterpreter() *Interpreter {
  env := NewEnv(nil)
  env.define("clock", &Clock{0})
  return &Interpreter{
    env       : env,
    statements: make([]Statement, 0),
  }
}

func (i *Interpreter) interpret (statements []Statement) error {
  for _, statement := range statements {
    err := statement.Accept(i)
    if err != nil {
      return err
    }
  }
  return nil
}

func (i *Interpreter) execute_statement(statement Statement) {
   statement.Accept(i)
}

func (i *Interpreter) visit_return_statement(statement *ReturnStatement) error {
  expr := i.evaluate(statement.val)
  panic(ReturnValue{expr})
}

func (i *Interpreter) visit_func_call_expression(statement *FuncCallExpression) (returnval interface{}){
  fun := i.env.get(statement.ident.Lexeme.(string))
  defer func() {
      if r := recover(); r != nil {
        returnval = r.(ReturnValue).value
        return 
      }
  }()
  i_copy := &Interpreter{
    env: i.env,
  }
  f_args := make([]interface{}, 0)
  for _, arg := range statement.args{
    expr := i.evaluate(arg)
    f_args = append(f_args, expr)
  }
  switch f := fun.(type) {
    case *CallableFunc:
        f.call(i_copy, f_args)
    case *Clock:
        f.call(i_copy, f_args)
    default:
        fmt.Println("Unsupported type")
  }
  return nil
}

func (i *Interpreter) visit_func_statement(statement *FunctionStatement){
  cf := &CallableFunc{
    declaration: statement,
  }
  i.env.define(statement.ident.Lexeme.(string), cf)
}

func (i *Interpreter) visit_for_statement(visitor *ForStatement){
  i.execute_statement(visitor.start)
  for i.evaluate(visitor.predicate) == true{
    i.execute_statement(visitor.inner_statements)
    i.evaluate(visitor.incre)
  }
}

func (i *Interpreter) visit_while_statement(visitor *WhileStatement){
  predicate := is_truthy(i.evaluate(visitor.predicate))
  for predicate {
    i.execute_statement(visitor.inner_statements)
    predicate = is_truthy(i.evaluate(visitor.predicate))
  }
}

func (i *Interpreter) visit_if_statement(visitor *IfStatement){
  predicate := is_truthy(i.evaluate(visitor.predicate))
  if predicate {
    i.execute_statement(visitor.if_condition)
  } else if !predicate && visitor.else_condition != nil{
    i.execute_statement(visitor.else_condition)
  }
}

func (i *Interpreter) visit_assign_expression(visitor *AssignExpression) interface{}{
  assigned := i.env.assign(visitor.ident.Lexeme.(string), i.evaluate(visitor.value))
  return assigned
}

func (i *Interpreter) visit_block_statement(visitor *BlockStatement) {
  env := NewEnv(i.env)
  prev := i.env
  i.env = env
  for _, statement := range visitor.statements{
    i.execute_statement(statement)
  }
  i.env = prev
}

func (i *Interpreter) execute_block(statement *BlockStatement, env *Env){
  prev := i.env
  i.env = env
  for _, statement := range statement.statements{
    i.execute_statement(statement)
  }
  i.env = prev
}

func (i *Interpreter) visit_debug_statement(visitor *DebugStatement){
  evaluated_expr := i.evaluate(visitor.expr)
  fmt.Println(evaluated_expr)
}

func (i *Interpreter) visit_var_statement(visitor *VarStatement){
  i.env.define(visitor.ident.Lexeme.(string), i.evaluate(visitor.value))
}

func (i *Interpreter) visit_logical_expression(visitor *LogicalExpression) interface{}{
  left := i.evaluate(visitor.left)
  right := i.evaluate(visitor.right)
  if visitor.op.Type == OR{
    return is_truthy(left) || is_truthy(right)
  }
  if visitor.op.Type == AND{
    return is_truthy(left) && is_truthy(right)
  }
  panic("Unreachable Code")
}

func (i *Interpreter) visit_expression_statement(visitor *ExpressionStatement){
  i.evaluate(visitor.expr)
}

func (i *Interpreter) evaluate(expr Expression) interface{} {
  return expr.Accept(i)
}

func (i *Interpreter) visit_binary_expression(visitor *BinaryExpression) interface{}{
  right := i.evaluate(visitor.right)
  left  := i.evaluate(visitor.left)
  rightType := reflect.TypeOf(right)
  leftType  := reflect.TypeOf(left)
  switch visitor.op.Type {
    case EQUAL_EQUAL: {
      return right == left
    }
    case BANG_EQUAL: {
      return right != left
    }
    case GREATER: {
      if rightType.Kind() == reflect.Float64 && leftType.Kind() == reflect.Float64 {
        return left.(float64) > right.(float64)
      }else if rightType.Kind() == reflect.String && leftType.Kind() == reflect.String {
        return left.(string) > right.(string)
      }
    }
    case GREATER_EQUAL: {
      if rightType.Kind() == reflect.Float64 && leftType.Kind() == reflect.Float64 {
        return left.(float64) >= right.(float64)
      }else if rightType.Kind() == reflect.String && leftType.Kind() == reflect.String {
        return left.(string) >= right.(string)
      }
    }
    case LESS: {
      if rightType.Kind() == reflect.Float64 && leftType.Kind() == reflect.Float64 {
        return left.(float64) < right.(float64)
      }else if rightType.Kind() == reflect.String && leftType.Kind() == reflect.String {
        return left.(string) < right.(string)
      }
    }
    case LESS_EQUAL: {
      if rightType.Kind() == reflect.Float64 && leftType.Kind() == reflect.Float64 {
        return left.(float64) <= right.(float64)
      }else if rightType.Kind() == reflect.String && leftType.Kind() == reflect.String {
        return left.(string) <= right.(string)
      }
    }
    case PLUS: {
      if rightType.Kind() == reflect.Float64 && leftType.Kind() == reflect.Float64 {
        return right.(float64) + left.(float64)
      }else if rightType.Kind() == reflect.String && leftType.Kind() == reflect.String {
        return right.(string) + left.(string)
      }
    }
    case MINUS: {
      if rightType.Kind() == reflect.Float64 && leftType.Kind() == reflect.Float64 {
        return left.(float64) - right.(float64)
      }
    }
    case STAR: {
      if rightType.Kind() == reflect.Float64 && leftType.Kind() == reflect.Float64 {
        return right.(float64) * left.(float64)
      }
    }
    case SLASH: {
      if rightType.Kind() == reflect.Float64 && leftType.Kind() == reflect.Float64 {
        return left.(float64) / right.(float64)
      }
    }
    default:
      fmt.Println(visitor.op.Type)
      panic("Undefined operator on binary expression")
  }
  return nil
}

func (i *Interpreter) visit_grouping_expression(visitor *GroupingExpression) interface{}{
  return i.evaluate(visitor.expr)
}

func (i *Interpreter) visit_literal_expression(visitor *LiteralExpression) interface{} {
  return visitor.value
}

func (i *Interpreter) visit_var_expression(visitor *VarExpression) interface{} {
  return i.env.get(visitor.ident.Lexeme.(string))
}

func (i *Interpreter) visit_unary_expression(visitor *UnaryExpression) interface {}{
  right := i.evaluate(visitor.right)
  switch v := right.(type){
    case int32, int64, float32, float64 : {
      if visitor.op.Type == MINUS{
        return v.(float64) * -1.0
      }
      if visitor.op.Type == BANG{
        return !is_truthy(v)
      }
    }
    case bool : {
      if visitor.op.Type == BANG{
        return !is_truthy(v)
      }
    }
    default:
      fmt.Println(v)
  }
  return nil
}

func is_truthy(value interface{}) bool{
  switch value.(type){
    case nil :
      return false
    case float64 :
      return true
    case string :
      return true
    case bool :
      return value.(bool)
    default:
      return true
  }
}

